Sprint 11

Kolesnikov-Eugene opened the pull request.

Kolesnikov-Eugene committed:
ProgressHUD added and SnakeCaseJSONDecoder

Kolesnikov-Eugene committed:
ProfileService added to project

Kolesnikov-Eugene committed:
URLRequest extension changed
Constants(added url for token and base url)

Kolesnikov-Eugene committed:
ProfileImageService, UserResult added

Kolesnikov-Eugene committed:
added observer for ProfileViewController and changed KVO for WebViewViewController

Kolesnikov-Eugene committed:
new API KVO for WKWebKit
func object with Generics
alert with error if failed

Kolesnikov-Eugene committed:
changed alert with error, works correctly now

Kolesnikov-Eugene committed:
added Kingfisher to load images and KeyChainWrapper to store token

Kolesnikov-Eugene committed:
changed AppDelegate and SceneDelegate
SplashScreen created programmatically

Kolesnikov-Eugene committed:
final ver
did some code refactoring

protos17 commented on FunImages/Auth/SnakeCaseJSONDecoder.swift:
```diff
@@ -0,0 +1,15 @@
+
+import Foundation
+
+class SnakeCaseJSONDecoder: JSONDecoder {
```
Придумано отлично! Но рекомендую его еще сделать final, потому что он не наследуется

protos17 commented on FunImages/Auth/URLRequest+Extensions.swift:
```diff
@@ -11,10 +11,15 @@ extension URLRequest {
         var request = URLRequest(url: URL(string: path, relativeTo: baseURL)!)
         request.httpMethod = httpMethod
+        if tokenNeededForRequest {
+            let token = OAuth2TokenStorage.shared.token!
```
Не рекомендую использовать принудительное извлечение опционалов, лучше добавить проверку через guard let

protos17 commented on FunImages/OAuth2/OAuth2Service.swift:
```diff
@@ -22,23 +24,48 @@ final class OAuth2Service {
             guard let self = self  else { return }
             switch result {
             case .success(let body):
+                self.task = nil
```
task рекомендую обнулять не только при успешном результате, а при любом

protos17 commented on FunImages/OAuth2/OAuth2TokenStorage.swift:
```diff
@@ -6,18 +6,21 @@
-            defaults.set(newValue, forKey: tokenKey)
+            if let newValue = newValue {
+                keyChainWrapper.set(newValue, forKey: tokenKey)
+            }
```
Рекомендую еще добавить метод удаления ключа, когда идет присваивание nil

protos17 commented on FunImages/Profile/ProfileImageService.swift:
```diff
@@ -0,0 +1,58 @@
+                        self.task = nil
+                        
+                        self.avatarURL = profilePhotoURL.profileImage.medium
+                        completion(.success(self.avatarURL!))
```
Рекомендую аккуратнее работать с опционалами здесь и в других местах. Принудительное извлечение опционалов может часто приводит к вылетам приложения

protos17 commented on FunImages/Profile/ProfileResult.swift:
```diff
@@ -0,0 +1,15 @@
+    let username: String
+    let firstName: String
+    let lastName: String
+    let bio: String
```
Чтобы правильно работала программа, необходимо сделать lastName и bio опциональными

protos17 commented on FunImages/Profile/ProfileService.swift:
```diff
@@ -0,0 +1,54 @@
+
+final class ProfileService {
+    static let shared = ProfileService()
+    var profileData: Profile?
```
Если свойство снаружи используется только для чтения, рекомендую указывать как private(set)

protos17 commented on FunImages/SplashView/File.swift:
Рекомендую удалить, лежит в файлах проекта

protos17 commented on FunImages/SplashView/SplashViewController.swift:
```diff
@@ -3,64 +3,138 @@
+    }
+
+    var token: String? {
+        oauth2Storage.token ?? nil
```
oauth2Storage.token уже имеет тип данных String?, нет необходимости указывать значение при отсутствии значения nil (то есть то же самое отсутствие значения))

protos17 commented on FunImages/SplashView/SplashViewController.swift:
```diff
@@ -3,64 +3,138 @@
-        if OAuth2TokenStorage.shared.token != nil {
-            switchToTabBarController()
+
+        if token != nil {
```
Здесь как раз пригодилось безопасное извлечение опционала:

if let token = token {
fetchProfile(token)
}

protos17 commented on FunImages/SplashView/SplashViewController.swift:
```diff
@@ -3,64 +3,138 @@
-            self.fetchAuthToken(code)
+            switch result {
+            case .success:
+                fetchProfile(token!)
```
Очень много в проекте принудительного извлечения опционалов

Kolesnikov-Eugene replied:
Пока работал над проектом, использовал принудительное извлечение, где был уверен, что значение будет. Понял, что далеко не лучшая практика) Лучше сразу использовать безопасное извлечение) Впредь буду аккуратнее)

protos17 commented on FunImages/SplashView/SplashViewController.swift:
```diff
@@ -3,64 +3,138 @@
+                    username: self.profileService.profileData!.userName) { result in
+                        switch result {
+                        case .success:
+                            print("success")
```
Printы обычно не приветствуются внутри основного кода приложения

protos17 commented on FunImages/SplashView/SplashViewController.swift:
```diff
@@ -3,64 +3,138 @@
+        let alert = UIAlertController(title: "Что-то пошло не так(",
+                                      message: "Не удалось войти в систему",
+                                      preferredStyle: .alert)
+        let action = UIAlertAction(title: "OK", style: .default) { [self] _ in
```
Рекомендую указать weak self

Kolesnikov-Eugene committed:
added safe unwrapping of Optionals in the project.
Refactored code in ProfileViewController(flow of initializing views).
lastName and bio in ProfileResult changed to Optionals to prevent crash of the app if user didnt fill out these fields in profile.

Kolesnikov-Eugene committed:
deleted File.swift

protos17 commented on FunImages/OAuth2/OAuth2TokenStorage.swift:
```diff
@@ -6,20 +6,27 @@
         }
     }
     
     private init() { }
```
Рекомендую указать сразу после shared

protos17 commented on FunImages/Profile/ProfileViewController.swift:
```diff
@@ -6,104 +6,150 @@
-    
-    private func setContentAndConstraintsToView() {
-        view.backgroundColor = UIColor(red: 0.102, green: 0.106, blue: 0.133, alpha: 1)
+    private var profileImageView: UIImageView = {
```
UI элементы рекомендую указывать как let или lazy var

protos17 commented on FunImages/Profile/ProfileViewController.swift:
```diff
@@ -6,104 +6,150 @@
-            profileImageView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 16)
-        ])
+    private func addSubViews() {
+        view.addSubview(profileImageView)
```
Можно указать так:

        [profileImageView,
         profileNameLabel ...].forEach { view.addSubview($0)}

protos17 commented on FunImages/SplashView/SplashViewController.swift:
```diff
@@ -3,64 +3,136 @@
             case .success:
+                UIBlockingProgressHUD.dismiss()
+                profileImageService.fetchProfileImageURL(
+                    username: self.profileService.profileData!.userName) { result in
```
ЗДесь тоже необходимо добавить weak self
